src/range.rs

```diff
@@ -0,0 +1,355 @@
+///
+/// # Translation from `text_unit`
+///
+/// - `TextRange::from_to(from, to)` ⟹ `TextRange::from(from..to)`
+/// - `TextRange::offset_len(offset, size)` ⟹ `TextRange::at(offset).with_len(size)`
+/// - `range.start()` ⟹ `range.start()`
+/// - `range.end()` ⟹ `range.end()`
+/// - `range.len()` ⟹ `range.len()`<sup>†</sup>
+/// - `range.is_empty()` ⟹ `range.is_empty()`
+/// - `a.is_subrange(b)` ⟹ `b.contains(a)`
+/// - `a.intersection(b)` ⟹ `TextRange::intersection(a, b)`
+/// - `a.extend_to(b)` ⟹ `TextRange::covering(a, b)`
+/// - `range.contains(offset)` ⟹ `range.contains_point(point)`
+/// - `range.contains_inclusive(offset)` ⟹ `range.contains_point_inclusive(point)`
+///
+/// † See the note on [`TextRange::len`] for differing behavior for incorrect reverse ranges.
+#[derive(Copy, Clone, Eq, PartialEq, Hash)]
+pub struct TextRange {
+    start: TextSize,
+    end: TextSize,
+}
+
+#[allow(non_snake_case)]
+pub(crate) const fn TextRange(start: TextSize, end: TextSize) -> TextRange {
+    TextRange { start, end }
+}
+
+impl fmt::Debug for TextRange {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        fmt::Display::fmt(self, f)
+    }
+}
+
+impl fmt::Display for TextRange {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "[{}..{})", self.start(), self.end())
+    }
+}
+
+/// Identity methods.
+impl TextRange {
+    /// The start point of this range.
+    pub const fn start(self) -> TextSize {
+        self.start
+    }
+
+    /// The end point of this range.
+    pub const fn end(self) -> TextSize {
+        self.end
+    }
+
+    /// The size of this range.
+    ///
+    /// # Panics
+    ///
+    /// When `end() < start()`, triggers a subtraction overflow.
+    /// This will panic with debug assertions, and overflow without.
+    pub const fn len(self) -> TextSize {
+        // HACK for const fn: math on primitives only
+        TextSize(self.end().raw - self.start().raw)
+    }
+
+    /// Check if this range empty or reversed.
+    ///
+    /// When `end() < start()`, this returns false.
+    /// Code should prefer `is_empty()` to `len() == 0`,
+    /// as this safeguards against incorrect reverse ranges.
+    pub const fn is_empty(self) -> bool {
+        // HACK for const fn: math on primitives only
+        self.start().raw >= self.end().raw
+    }
+}
+
+/// Manipulation methods.
+impl TextRange {
+    /// A range covering the text size of some text-like object.
+    pub fn of(size: impl TextSized) -> TextRange {
+        TextRange(0.into(), size.text_size())
+    }
+
+    /// An empty range at some text size offset.
+    pub fn at(size: impl Into<TextSize>) -> TextRange {
+        let size = size.into();
+        TextRange(size, size)
+    }
+
+    /// Set the length without changing the starting offset.
+    pub fn with_len(self, len: impl Into<TextSize>) -> TextRange {
+        TextRange(self.start(), self.start() + len.into())
+    }
```

Comment:
I think we should provide, two, and only two, constructing functions: from start and len, and from start and end. I am not sure what's the best shape for those would be. I don't have text_unit's current `from_to`, `offset_len`. I briefly thought about `TextRange(start..end)` and `TextRange::offset_len(start, len)`. Heck, right now I am thinking of `start.range_to(end)` / `start.range_of(len)`.

I suggest starting with a conservative option of two factory functions, and leave the bikesheding for a more fluent API until later.

Reply:
>For the immediate, I'm punting on the "start + length" constructor; this can be re-added after this PR with a dedicated bikeshed thread stuck_out_tongue

:+1: 

I've also noticed that a few instalce of `start +length` ctor in rust-analyzer can be replaced by `TextRange::empty(offset: TextSize)`.